control-plane-operator/controllers/hostedcontrolplane/cloud/powervs/powervs.go

```diff
@@ -0,0 +1,207 @@
+package powervs
+
+import (
+	"bytes"
+	"fmt"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/releaseinfo"
+	"strings"
+	"text/template"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+const (
+	ccmContainerName       = "cloud-controller-manager"
+	kubeConfigFileBasePath = "/etc/kubernetes"
+	secretMountPath        = "/etc/vpc"
+	ccmConfigMapMountPath  = "/etc/ibm"
+)
+
+var ccmConfigTemplateData = `
+[global]
+version = 1.1.0
+[kubernetes]
+config-file = {{.ConfigFile}}
+[provider]
+cluster-default-provider = g2
+accountID = {{.AccountID}}
+clusterID = {{.ClusterID}}
+g2workerServiceAccountID = {{.G2workerServiceAccountID}}
+g2Credentials = {{.G2Credentials}}
+g2ResourceGroupName = {{.G2ResourceGroupName}}
+g2VpcSubnetNames = {{.G2VpcSubnetNames}}
+g2VpcName = {{.G2VpcName}}
+region = {{.Region}}
+powerVSCloudInstanceID = {{.PowerVSCloudInstanceID}}
+powerVSRegion = {{.PowerVSRegion}}
+powerVSZone = {{.PowerVSZone}}`
+
+var ccmConfigTemplate = template.Must(template.New("ccmConfigMap").Parse(ccmConfigTemplateData))
+
+func ReconcileCCMConfigMap(ccmConfig *corev1.ConfigMap, hcp *hyperv1.HostedControlPlane) (err error) {
+	config := map[string]string{
+		"ConfigFile":               fmt.Sprintf("%s/kubeconfig", kubeConfigFileBasePath),
+		"AccountID":                hcp.Spec.Platform.PowerVS.AccountID,
+		"ClusterID":                hcp.Name,
+		"G2workerServiceAccountID": hcp.Spec.Platform.PowerVS.AccountID,
+		"G2Credentials":            fmt.Sprintf("%s/%s", secretMountPath, "ibmcloud_api_key"),
+		"G2ResourceGroupName":      hcp.Spec.Platform.PowerVS.ResourceGroup,
+		"G2VpcSubnetNames":         hcp.Spec.Platform.PowerVS.VPC.Subnet,
+		"G2VpcName":                hcp.Spec.Platform.PowerVS.VPC.Name,
+		"Region":                   hcp.Spec.Platform.PowerVS.VPC.Region,
+		"PowerVSCloudInstanceID":   hcp.Spec.Platform.PowerVS.ServiceInstanceID,
+		"PowerVSRegion":            hcp.Spec.Platform.PowerVS.Region,
+		"PowerVSZone":              hcp.Spec.Platform.PowerVS.Zone,
+	}
+
+	configData := &bytes.Buffer{}
+	ccmConfigTemplate.Execute(configData, config)
+
+	if ccmConfig.Data == nil {
+		ccmConfig.Data = map[string]string{}
+	}
+
+	ccmConfig.Data[ccmConfig.Name] = configData.String()
+
+	return
+}
+
+func ReconcileCCMDeployment(deployment *appsv1.Deployment, hcp *hyperv1.HostedControlPlane, ccmConfig *corev1.ConfigMap, releaseImage *releaseinfo.ReleaseImage) (err error) {
+	commandToExec := []string{
+		"/bin/ibm-cloud-controller-manager",
```

Comment:
The Machine config server will set the external flag for kube appropriately based on the platform https://github.com/openshift/machine-config-operator/blob/1af0a6f96fc7c0f44230293ac026a96acb0da512/pkg/controller/template/render.go#L373-L390
However don't we need to ensure in hypershift for other components e.g KCM/KAS they don't set the cloud provider flag https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager?

Reply:
Make sense, we should have some function which decides what's the option for any given platform, how do we deal with an provider like aws which has both inTree / OutOfTree? want to set something suitable default mode in the code or give user an option to select.?

Reply:
Let's consider this an impl detail for now, don't give the user a choice. Your are the first one out of tree so we'll need to flesh it out here, the others are in-tree for now.

Reply:
Both KAS and KCM does not have `--cloud-provider` flag not configured by default. 
Based on the platform, CloudProvider param is configured and later `--cloud-provider` flag is set. 
For OutOfTree cloud-provider, we can avoid setting CloudProvider param which results in not setting `--cloud-provider` flag.

KAS CloudProvider param: https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/kas/params.go#L284
KCM CloudProvider param: https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/kcm/params.go#L102

`--cloud-provider` flag: https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/kcm/deployment.go#L305

Hope, this is fine.